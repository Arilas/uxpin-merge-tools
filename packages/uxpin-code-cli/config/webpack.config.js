module.exports = {
  entry: './src/steps/styleGuide/App.js',
  output: {
    filename: './dist/styleguide.js',
  },
  resolve: {
    extensions: ['.js', '.jsx', '.json'],
  },
};
