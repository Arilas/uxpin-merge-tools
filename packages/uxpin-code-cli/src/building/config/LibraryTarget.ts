export type LibraryTarget = 'amd'|'commonjs';
