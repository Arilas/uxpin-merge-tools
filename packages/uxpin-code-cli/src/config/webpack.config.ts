import { BabelPlugin } from '../building/plugins/BabelPluginDefinition';

const LOADER_BABEL:string = 'babel-loader';

const CONFIG:any = {
  entry: './src/components.js',
  module: {
    rules: [
      {
        test: /\.jsx?$/,
        use: [{
          loader: 'babel-loader',
          options: {
            babelrc: false,
            plugins: [
            ],
            presets: [
              'react',
              ['env', {
                targets: {
                  browsers: ['last 2 versions'],
                },
              }],
              'stage-0',
            ],
          },
        }],
      },
    ],
  },
  output: {
    filename: 'components.js',
    libraryTarget: 'amd',
  },
  resolve: {
    extensions: ['.js', '.jsx'],
  },
};

export function getConfig(babelPlugins:BabelPlugin[], target?:string):any {
  return getConfigDecoratedWithLibraries(babelPlugins, target);
}

function getConfigDecoratedWithLibraries(babelPlugins:BabelPlugin[], target?:string):any {
  const config:any = Object.assign({}, CONFIG);

  if (target) {
    config.output.libraryTarget = target;
  }

  config.module.rules.forEach((rule:any) => {
    if (!rule || !rule.use) {
      return;
    }

    rule.use.forEach((loader:any) => {
      if (loader.loader === LOADER_BABEL) {
        loader.options.plugins = babelPlugins;
      }
    });
  });

  return config;
}
