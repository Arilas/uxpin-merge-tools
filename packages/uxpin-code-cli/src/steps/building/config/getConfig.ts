import { isEmpty } from 'lodash';
import { join, relative } from 'path';
import { Configuration } from 'webpack';
import { smart } from 'webpack-merge';

export const TEMP_DIR_PATH:string = './.uxpin-temp';
export const LIBRARY_INPUT_PATH:string = `${TEMP_DIR_PATH}/components.js`;
export const LIBRARY_OUTPUT_PATH:string = `${TEMP_DIR_PATH}/designsystemlibrary.js`;

export function getConfig(webpackConfigPath:string = ''):Configuration {
  const config:Configuration = {
    entry: LIBRARY_INPUT_PATH,
    output: {
      filename: LIBRARY_OUTPUT_PATH,
      libraryTarget: 'commonjs',
    },
    resolve: {
      extensions: ['.js', '.jsx'],
    },
  };

  if (isEmpty(webpackConfigPath)) {
    return config;
  }

  const path:string = relative(__dirname, join(process.cwd(), webpackConfigPath));
  const userWebpackConfig:Configuration = require(path);
  return smart(userWebpackConfig, config);
}
