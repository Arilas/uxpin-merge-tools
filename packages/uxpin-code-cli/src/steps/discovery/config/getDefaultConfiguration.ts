import { CliConfig } from './CliConfig';

export function getDefaultConfiguration():CliConfig {
  return {
    components: {
      categories: [
        {
          include: [
            'src/components/*/*.{js,jsx,ts,tsx}',
            'src/*/*.{js,jsx,ts,tsx}',
          ],
          name: 'Uncategorized',
        },
      ],
    },
  };
}
