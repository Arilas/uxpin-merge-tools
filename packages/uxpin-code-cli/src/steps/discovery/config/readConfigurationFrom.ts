import { join } from 'path';
import { CliConfig } from './CliConfig';

const CONFIG_FILE_NAME:string = 'uxpin.config.js';

export function readConfigurationFrom(directoryPath:string):CliConfig | undefined {
  try {
    return require(join(directoryPath, CONFIG_FILE_NAME));
  } catch (e) {
    console.log('Error while reading configuration file. Using default configuration');
    console.log(e);
    return;
  }
}
