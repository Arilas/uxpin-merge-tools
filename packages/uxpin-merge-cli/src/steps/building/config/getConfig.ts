import * as path from 'path';
import { Configuration } from 'webpack';
import { merge as webpackMerge } from 'webpack-merge';
import { BuildOptions } from '../BuildOptions';
import { getComponentLibraryInputPath } from '../library/getComponentLibraryInputPath';

export const DEBUG_APP_BUNDLED_FILE:string = 'index.js';
export const TEMP_DIR_NAME:string = '.uxpin-merge';
export const TEMP_DIR_PATH:string = `./${TEMP_DIR_NAME}`;
export const LIBRARY_INPUT_FILENAME:string = `components.js`;
export const LIBRARY_OUTPUT_FILENAME:string = 'designsystemlibrary.js';

// Storybook generated webpack related variables
export const SB_WEBPACK_JS_LOADER_TEST:string = '/\\.(js|mjs|jsx|ts|tsx)$/';
export const SB_WEBPACK_BABEL_LOADER_SUFFIX:string = 'babel-loader/lib/index.js';

// Check whether an object that looks like a generated webpack loader configuration
// looks like something meant to load .js/.ts files via babel-loader
function isJSWebpackLoader(o:any):boolean {
  return typeof o === 'object'
    && 'loader' in o && o.loader && Array.isArray(o.loader)
    && o.loader.includes(SB_WEBPACK_BABEL_LOADER_SUFFIX)
    && 'test' in o && typeof o.test === 'string'
    && o.test === SB_WEBPACK_JS_LOADER_TEST;
}

export function getConfig(
  {
    development,
    webpackConfigPath,
    projectRoot,
    uxpinDirPath,
    storybook,
    storybookWebpackConfigPath,
  }:BuildOptions,
):Configuration {

  let config:Configuration = decorateWithDevToolsWhenDevelopment({
    entry: getComponentLibraryInputPath(uxpinDirPath),
    mode: development ? 'development' : 'production',
    optimization: {
      runtimeChunk: false,
      splitChunks: false,
    },
    output: {
      filename: LIBRARY_OUTPUT_FILENAME,
      libraryTarget: 'commonjs',
      path: uxpinDirPath,
    },
    resolve: {
      extensions: ['.js', '.jsx'],
    },
  }, development);

  // Join the webpack configuration for the project with the config defined above
  if (webpackConfigPath) {
    const userWebpackConfig:Configuration = require(path.join(projectRoot, webpackConfigPath));
    config = webpackMerge(userWebpackConfig, config);
  }

  // Add storybook configuration if storybook is enabled
  if (storybook && storybookWebpackConfigPath) {
    const storybookWebpackConfig:any = require(path.join(projectRoot, storybookWebpackConfigPath));

    // Look through the storybook webpack config and remove any loaders that are babel-loaders
    // storybook's loaders interfere with the loading done for js and ts files locally and cause the
    // exported bundle to not have any static exports (and be much shorter than normal)
    if (storybookWebpackConfig.module && storybookWebpackConfig.module.rules) {
      storybookWebpackConfig.module.rules = storybookWebpackConfig.module.rules.filter((r:any) => {
        // Filter out any webpack loaders that storybook is trying to use
        if (isJSWebpackLoader(r)) {
          return false;
        }

        // If the loader rule is not a oneOf (and is not for js) it can be included
        if (!r.oneOf) { return true; }

        // Do the same if dealing with a oneOf rule (which is an aggregate)
        r.oneOf = r.oneOf.filter((o:any) => {
          // Remove js loaders in 'oneOf' composite loaders
          if (isJSWebpackLoader(o)) {
            return false;
          }

          return true;
        });
      });
    }

    // Merge the storybook webpack config with the app config
    config = webpackMerge(storybookWebpackConfig, config);
  }

  return config;
}

function decorateWithDevToolsWhenDevelopment(config:Configuration, development:boolean = false):Configuration {
  if (!development) {
    return config;
  }

  return {
    ...config,
    devtool: 'cheap-module-eval-source-map',
  };
}
