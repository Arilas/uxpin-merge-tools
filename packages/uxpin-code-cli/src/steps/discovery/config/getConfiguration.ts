import { CliConfig } from './CliConfig';
import { getDefaultConfiguration } from './getDefaultConfiguration';
import { readConfigurationFrom } from './readConfigurationFrom';

export function getConfiguration(projectRoot:string):CliConfig {
  return readConfigurationFrom(projectRoot) || getDefaultConfiguration();
}
