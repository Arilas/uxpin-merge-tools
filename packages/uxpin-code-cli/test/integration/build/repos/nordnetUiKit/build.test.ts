import { relative, resolve } from 'path';

import { LIBRARY_OUTPUT_PATH } from '../../../../../src/building/config/getConfig';
import { runUXPinCodeCommand } from '../../../../utils/command/runUXPinCodeCommand';
import { setTimeoutBeforeAll } from '../../../../utils/command/setTimeoutBeforeAll';

const CURRENT_TIMEOUT:number = 75000;

setTimeoutBeforeAll(CURRENT_TIMEOUT);

describe('Building repos/nordnet-ui-kit design system', () => {
  describe('with required user webpack config', () => {
    let components:any;
    let consoleOutput:string;

    beforeAll(() => {
      const options:string = [
        '--webpack-config "../../configs/nordnet-ui-kit-webpack.config.js"',
        '--wrapper "../documentation/wrapper.jsx"',
      ].join(' ');

      return runUXPinCodeCommand('resources/repos/nordnet-ui-kit', options)
        .then((output) => {
          const path:string = relative(__dirname, resolve('test/resources/repos/nordnet-ui-kit', LIBRARY_OUTPUT_PATH));
          components = require(path);
          consoleOutput = output;
        });
    });

    it('contains Button component', () => {
      // when
      const { Button } = components;
      // then
      expect(Button).toBeInstanceOf(Function);
    });

    it('contains custom wrapper', () => {
      // when
      const { Wrapper } = components;
      // then
      expect(Wrapper).toBeInstanceOf(Function);
    });

    it('prints warnings without stack traces to the console', () => {
      // then
      expect(consoleOutput).toMatchSnapshot();
    });
  });

  describe('without required user webpack config', () => {
    it('throws an error', () => {
      return runUXPinCodeCommand('resources/repos/nordnet-ui-kit')
        .then((output) => {
          expect(output).toContain('ERROR:');
          expect(output).toContain('Module parse failed');
        });
    });
  });
});
