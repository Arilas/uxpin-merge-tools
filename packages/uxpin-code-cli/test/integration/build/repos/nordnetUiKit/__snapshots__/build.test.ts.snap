// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Building repos/nordnet-ui-kit design system with required user webpack config prints warnings without stack traces to the console 1`] = `
"Alert
    ðŸ“œ documentation: âœ”

Animate
    ðŸ“œ documentation: âœ”

Avatar
    ðŸ“œ documentation: âœ”

Badge
    ðŸ“œ documentation: âœ”

Button
    ðŸ“œ documentation: âœ”

Dropdown
    ðŸ“œ documentation: âœ”

Flag
    ðŸ“œ documentation: âœ”

Icon
    ðŸ“œ documentation: âœ”

Input
    ðŸ“œ documentation: âœ”

LabeledValue
    ðŸ“œ documentation: âœ”

Li
    ðŸ“œ documentation: âœ”

Logo
    ðŸ“œ documentation: âœ”

Pane
    ðŸ“œ documentation: âœ”

RadioGroup
    ðŸ“œ documentation: âœ”

SegmentedControl
    ðŸ“œ documentation: âœ”

Spinner
    ðŸ“œ documentation: âœ”

Table
    ðŸ“œ documentation: âœ”

Tbody
    ðŸ“œ documentation: âœ˜

Td
    ðŸ“œ documentation: âœ˜

Tfoot
    ðŸ“œ documentation: âœ˜

Th
    ðŸ“œ documentation: âœ˜

Thead
    ðŸ“œ documentation: âœ˜

Tooltip
    ðŸ“œ documentation: âœ”

Tr
    ðŸ“œ documentation: âœ˜

Ul
    ðŸ“œ documentation: âœ”

warning Cannot serialize component properties in:
src/components/alert/alert.jsx

warning Cannot serialize component properties in:
src/components/animate/animate.jsx

warning Cannot serialize component properties in:
src/components/avatar/avatar.jsx

warning Cannot serialize component properties in:
src/components/badge/badge.jsx

warning Cannot serialize component properties in:
src/components/button/button.jsx

warning Cannot serialize component properties in:
src/components/dropdown/dropdown.jsx

warning Cannot parse type of a property 'aria-hidden' in:
src/components/icon/icon.jsx

warning Cannot parse type of a property 'focusable' in:
src/components/icon/icon.jsx

warning Cannot serialize component properties in:
src/components/labeled-value/labeled-value.jsx

warning Cannot serialize component properties in:
src/components/li/li.jsx

warning Cannot serialize component properties in:
src/components/pane/pane.jsx

warning Cannot serialize component properties in:
src/components/spinner/spinner.jsx

warning Cannot serialize component properties in:
src/components/td/td.jsx

warning Cannot serialize component properties in:
src/components/th/th.jsx

warning Cannot serialize component properties in:
src/components/tooltip/tooltip.jsx

"
`;
